File: INFRA/style/themes/r/eig_theme.R
# EIG plotting/table theme helpers for R.

eig_load_tokens <- function(path = "INFRA/style/themes/r/eig_tokens.R") {
  if (!file.exists(path)) {
    stop("Token file not found: ", path, call. = FALSE)
  }
  env <- new.env(parent = baseenv())
  sys.source(path, envir = env)
  env
}

eig_assert_fonts <- function(tokens = NULL, allow_fallback = FALSE) {
  if (is.null(tokens)) tokens <- eig_load_tokens()
  if (requireNamespace("systemfonts", quietly = TRUE)) {
    available <- unique(systemfonts::system_fonts()$family)
  } else {
    fc <- suppressWarnings(system("fc-list : family", intern = TRUE))
    if (length(fc) == 0) {
      stop(
        "Unable to enumerate system fonts (install 'systemfonts' or provide fontconfig).",
        call. = FALSE
      )
    }
    parts <- trimws(unlist(strsplit(fc, ",")))
    available <- unique(parts[nchar(parts) > 0])
  }
  headline_primary <- tokens$EIG_FONT_HEADLINE_PRIMARY
  body_primary <- tokens$EIG_FONT_BODY_PRIMARY

  if (!allow_fallback) {
    missing <- c()
    # Treat Tiempos Text as strict-equivalent for Tiempos Headline in modern installs.
    headline_ok <- headline_primary %in% available ||
      (identical(headline_primary, "Tiempos Headline") && "Tiempos Text" %in% available)
    if (!headline_ok) missing <- c(missing, headline_primary)
    # macOS variable-font installs may expose Galaxie Polaris as "OpenSans".
    body_ok <- body_primary %in% available ||
      (identical(body_primary, "Galaxie Polaris") && "OpenSans" %in% available)
    if (!body_ok) missing <- c(missing, body_primary)
    if (length(missing) > 0) {
      stop(
        "Missing required EIG primary fonts: ",
        paste(missing, collapse = ", "),
        ". Run scripts/fonts installers and restart session.",
        call. = FALSE
      )
    }
    return(invisible(TRUE))
  }

  headline_stack <- c(headline_primary, "Tiempos Text", "Source Serif 3", "Georgia", "Times New Roman")
  body_stack <- c(body_primary, "Arial", "Helvetica Neue", "Helvetica")
  if (!any(headline_stack %in% available)) stop("No valid headline font available.", call. = FALSE)
  if (!any(body_stack %in% available)) stop("No valid body font available.", call. = FALSE)
  invisible(TRUE)
}

eig_theme_ggplot <- function(tokens = NULL, base_size = 10) {
  if (is.null(tokens)) tokens <- eig_load_tokens()
  if (!requireNamespace("ggplot2", quietly = TRUE)) {
    stop("Package 'ggplot2' is required.", call. = FALSE)
  }

  ggplot2::theme_minimal(base_size = base_size, base_family = tokens$EIG_FONT_BODY_PRIMARY) +
    ggplot2::theme(
      plot.title = ggplot2::element_text(
        family = tokens$EIG_FONT_HEADLINE_PRIMARY,
        face = "bold",
        size = base_size * 1.5,
        color = tokens$EIG_COLORS[["eig_black"]],
        hjust = 0
      ),
      plot.subtitle = ggplot2::element_text(
        family = tokens$EIG_FONT_BODY_PRIMARY,
        face = "bold",
        size = base_size * 1.1,
        color = tokens$EIG_COLORS[["eig_teal_900"]],
        hjust = 0
      ),
      axis.title = ggplot2::element_text(face = "bold", color = tokens$EIG_COLORS[["eig_black"]]),
      axis.text = ggplot2::element_text(color = tokens$EIG_COLORS[["eig_black"]]),
      legend.title = ggplot2::element_text(face = "bold"),
      panel.grid.minor = ggplot2::element_blank(),
      panel.grid.major.x = ggplot2::element_blank(),
      panel.grid.major.y = ggplot2::element_line(color = "#D9D9D9", linewidth = 0.3),
      plot.background = ggplot2::element_rect(fill = "white", color = NA),
      panel.background = ggplot2::element_rect(fill = "white", color = NA),
      plot.caption = ggplot2::element_text(size = base_size * 0.8, color = "#444444", hjust = 0)
    )
}

eig_scale_color <- function(tokens = NULL, ...) {
  if (is.null(tokens)) tokens <- eig_load_tokens()
  if (!requireNamespace("ggplot2", quietly = TRUE)) {
    stop("Package 'ggplot2' is required.", call. = FALSE)
  }
  ggplot2::scale_color_manual(values = tokens$EIG_DISCRETE, ...)
}

eig_scale_fill <- function(tokens = NULL, ...) {
  if (is.null(tokens)) tokens <- eig_load_tokens()
  if (!requireNamespace("ggplot2", quietly = TRUE)) {
    stop("Package 'ggplot2' is required.", call. = FALSE)
  }
  ggplot2::scale_fill_manual(values = tokens$EIG_DISCRETE, ...)
}

eig_scale_fill_seq <- function(which = c("eig_seq_blue_5", "eig_seq_red_5"), tokens = NULL, ...) {
  if (is.null(tokens)) tokens <- eig_load_tokens()
  if (!requireNamespace("ggplot2", quietly = TRUE)) {
    stop("Package 'ggplot2' is required.", call. = FALSE)
  }
  which <- match.arg(which)
  cols <- unname(unlist(tokens$EIG_SEQUENTIAL[[which]]))
  ggplot2::scale_fill_gradientn(colors = cols, ...)
}

eig_base_defaults <- function(tokens = NULL) {
  if (is.null(tokens)) tokens <- eig_load_tokens()
  old <- graphics::par(
    family = tokens$EIG_FONT_BODY_PRIMARY,
    bg = "white",
    fg = tokens$EIG_COLORS[["eig_black"]],
    col.axis = tokens$EIG_COLORS[["eig_black"]],
    col.lab = tokens$EIG_COLORS[["eig_black"]],
    las = 1,
    bty = "l"
  )
  invisible(old)
}

eig_plotly_layout <- function(tokens = NULL) {
  if (is.null(tokens)) tokens <- eig_load_tokens()
  list(
    font = list(family = tokens$EIG_FONT_BODY_PRIMARY, size = 11, color = tokens$EIG_COLORS[["eig_black"]]),
    title = list(font = list(family = tokens$EIG_FONT_HEADLINE_PRIMARY, size = 20, color = tokens$EIG_COLORS[["eig_black"]])),
    paper_bgcolor = tokens$EIG_COLORS[["eig_white"]],
    plot_bgcolor = tokens$EIG_COLORS[["eig_white"]],
    colorway = unname(unlist(tokens$EIG_DISCRETE)),
    xaxis = list(showgrid = FALSE, zeroline = FALSE),
    yaxis = list(showgrid = TRUE, gridcolor = "#D9D9D9", zeroline = FALSE)
  )
}

# Save a ggplot figure using EIG defaults.
# For PDF output, uses cairo_pdf device (not the string "pdf") to support EIG
# custom fonts (Tiempos Text, Galaxie Polaris) on all platforms. Passing
# device = "pdf" uses the base R PDF device and throws "invalid font type"
# with non-system fonts.
eig_save_fig <- function(plot, path, width = 10, height = 6, dpi = 300) {
  if (!requireNamespace("ggplot2", quietly = TRUE)) {
    stop("Package 'ggplot2' is required.", call. = FALSE)
  }
  if (grepl("\\.pdf$", path, ignore.case = TRUE)) {
    ggplot2::ggsave(path, plot = plot, width = width, height = height,
                   dpi = dpi, device = cairo_pdf)
  } else {
    ggplot2::ggsave(path, plot = plot, width = width, height = height, dpi = dpi)
  }
  invisible(path)
}

eig_style_gt <- function(gt_tbl, tokens = NULL) {
  if (is.null(tokens)) tokens <- eig_load_tokens()
  if (!requireNamespace("gt", quietly = TRUE)) {
    stop("Package 'gt' is required.", call. = FALSE)
  }

  gt_tbl |>
    gt::tab_options(
      table.border.top.width = gt::px(0),
      table.border.bottom.width = gt::px(0),
      heading.border.bottom.width = gt::px(0),
      column_labels.border.top.width = gt::px(0),
      column_labels.border.bottom.color = "#D9D9D9",
      table_body.hlines.color = "#E6E6E6",
      table_body.hlines.width = gt::px(1)
    ) |>
    gt::tab_style(
      style = list(
        gt::cell_fill(color = tokens$EIG_COLORS[["eig_teal_900"]]),
        gt::cell_text(color = "white", weight = "bold", font = tokens$EIG_FONT_BODY_PRIMARY)
      ),
      locations = gt::cells_column_labels(columns = gt::everything())
    )
}

File: Infrastructure/style/themes/python/eig_theme.py
"""EIG plotting/table theme helpers for Python."""

from __future__ import annotations

from typing import Iterable

import matplotlib.font_manager as fm
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
import plotly.graph_objects as go

try:
    from .eig_tokens import (
        BRAND_COLORS,
        EIG_DISCRETE,
        LEGACY_SEMANTIC_SETS,
        SEQUENTIAL,
        TYPOGRAPHY,
        SIZING,
    )
except ImportError:
    from eig_tokens import (  # type: ignore
        BRAND_COLORS,
        EIG_DISCRETE,
        LEGACY_SEMANTIC_SETS,
        SEQUENTIAL,
        TYPOGRAPHY,
        SIZING,
    )


def _installed_fonts() -> set[str]:
    return {f.name for f in fm.fontManager.ttflist if getattr(f, "name", None)}


def assert_eig_fonts(allow_fallback: bool = False) -> None:
    """Raise RuntimeError when required fonts are not available."""
    available = _installed_fonts()
    headline_primary = TYPOGRAPHY["headline"]["primary_family"]
    body_primary = TYPOGRAPHY["body"]["primary_family"]

    if not allow_fallback:
        missing = [x for x in (headline_primary, body_primary) if x not in available]
        if missing:
            raise RuntimeError(
                "Missing required EIG primary fonts: "
                + ", ".join(missing)
                + ". Run scripts/fonts installers and restart session."
            )
        return

    headline_stack = [headline_primary] + TYPOGRAPHY["headline"]["fallback_stack"]
    body_stack = [body_primary] + TYPOGRAPHY["body"]["fallback_stack"]
    if not any(x in available for x in headline_stack):
        raise RuntimeError(f"No available headline font in stack: {headline_stack}")
    if not any(x in available for x in body_stack):
        raise RuntimeError(f"No available body font in stack: {body_stack}")


def eig_palette(name: str = "discrete") -> list[str]:
    if name == "discrete":
        return list(EIG_DISCRETE)
    if name in SEQUENTIAL:
        return list(SEQUENTIAL[name])
    if name in LEGACY_SEMANTIC_SETS:
        return list(LEGACY_SEMANTIC_SETS[name].values())
    raise KeyError(f"Unknown EIG palette: {name}")


def set_eig_theme(figsize: tuple[float, float] | None = None) -> None:
    if figsize is None:
        fig = SIZING["figure_default"]
        figsize = (float(fig["width_in"]), float(fig["height_in"]))

    plt.rcParams.update(
        {
            "figure.figsize": figsize,
            "figure.facecolor": BRAND_COLORS["eig_white"],
            "axes.facecolor": BRAND_COLORS["eig_white"],
            "axes.edgecolor": "#CFCFCF",
            "axes.labelcolor": BRAND_COLORS["eig_black"],
            "axes.titlecolor": BRAND_COLORS["eig_black"],
            "axes.titlesize": 15,
            "axes.titleweight": "semibold",
            "axes.titlelocation": "left",
            "xtick.color": BRAND_COLORS["eig_black"],
            "ytick.color": BRAND_COLORS["eig_black"],
            "grid.color": "#D9D9D9",
            "grid.linewidth": 0.6,
            "font.family": TYPOGRAPHY["body"]["primary_family"],
            "text.color": BRAND_COLORS["eig_black"],
            "legend.frameon": False,
        }
    )
    sns.set_palette(EIG_DISCRETE)
    sns.set_style("whitegrid", {"axes.grid.axis": "y", "axes.grid": True})


def eig_plotly_template() -> go.layout.Template:
    return go.layout.Template(
        layout=dict(
            font=dict(
                family=TYPOGRAPHY["body"]["primary_family"],
                size=11,
                color=BRAND_COLORS["eig_black"],
            ),
            title=dict(
                font=dict(
                    family=TYPOGRAPHY["headline"]["primary_family"],
                    size=20,
                    color=BRAND_COLORS["eig_black"],
                ),
                x=0,
            ),
            colorway=list(EIG_DISCRETE),
            paper_bgcolor=BRAND_COLORS["eig_white"],
            plot_bgcolor=BRAND_COLORS["eig_white"],
            xaxis=dict(showgrid=False, zeroline=False),
            yaxis=dict(showgrid=True, gridcolor="#D9D9D9", zeroline=False),
            legend=dict(bgcolor="rgba(0,0,0,0)"),
        )
    )


def style_table(df: pd.DataFrame) -> pd.io.formats.style.Styler:
    return df.style.set_table_styles(
        [
            {
                "selector": "th",
                "props": [
                    ("background-color", BRAND_COLORS["eig_teal_900"]),
                    ("color", BRAND_COLORS["eig_white"]),
                    ("font-family", TYPOGRAPHY["body"]["primary_family"]),
                    ("font-size", "10pt"),
                    ("font-weight", "600"),
                    ("text-align", "left"),
                    ("border", "0"),
                ],
            },
            {
                "selector": "td",
                "props": [
                    ("font-family", TYPOGRAPHY["body"]["primary_family"]),
                    ("font-size", "9pt"),
                    ("color", BRAND_COLORS["eig_black"]),
                    ("border-top", "1px dotted #D9D9D9"),
                    ("border-bottom", "1px dotted #D9D9D9"),
                    ("border-left", "0"),
                    ("border-right", "0"),
                ],
            },
            {"selector": "table", "props": [("border-collapse", "collapse")]},
        ]
    )

